create_commits.py

```diff
@@ -0,0 +1,347 @@
+        remote_branch_name (str): The name of the branch.
+        expected_head_oid (str): The expected git commit oid at the head of the branch prior to the commit.
+        file_changes (dict): The file changes object.
+        message (str): The commit message.
+
+    Returns:
+        str: The commit oid of the created commit.
+    """
+
+    assert not remote_branch_name.startswith(
+        "origin/"
+    ), "Do not include 'origin/' in the remote branch name."
+
+    assert not remote_branch_name.startswith(
+        "refs/heads/"
+    ), "Do not include 'refs/heads/' in the remote branch name."
+
+    url = "https://api.github.com/graphql"
+    headers = {"Authorization": f"Bearer {github_token}"}
+
+    mutation = """
+    mutation ($input: CreateCommitOnBranchInput!) {
+      createCommitOnBranch(input: $input) {
+        commit {
+          oid
+        }
+      }
+    }
+    """
+
+    graphql_input = {
+        "branch": {
+            "repositoryNameWithOwner": repository_name_with_owner,
+            "branchName": remote_branch_name,
+        },
+        "expectedHeadOid": expected_head_oid,
+        "fileChanges": file_changes,
+        "message": message,
+    }
+
+    data = {"query": mutation, "variables": {"input": graphql_input}}
+    response = requests.post(url, headers=headers, json=data)
+
+    if "errors" in response.json():
+        logging.error(response.json())
+        exit(1)
+
+    return response.json()["data"]["createCommitOnBranch"]["commit"]["oid"]
+
+
+def main(
+    *,
+    github_token: str,
+    repository_name_with_owner: str,
+    local_branch_name: str,
+    remote_branch_name: str,
+) -> None:
+    """
+    Create commits on a remote branch for each commit on the local branch that's not on the remote branch.
+
+    Note: This function assumes that the commits created on the remote branch are unlikely to have
+    the same commit ID as the local commits. This is intentional - the createCommitOnBranch mutation
+    handles commit signing and commit authorship attribution for us, so the commit contents will be
+    different than the local commits. As a result, the commit IDs/hashes will be different as well.
+
+    Args:
+        github_token (str): The GitHub personal access github_token.
+        repository_name_with_owner (str): The name of the repository with the owner.
+        local_branch_name (str): The name of the local branch.
+        remote_branch_name (str): The name of the remote branch.
+    """
+    assert not remote_branch_name.startswith(
+        "origin/"
+    ), "Do not include 'origin/' in the remote branch name."
+    assert not remote_branch_name.startswith(
+        "refs/heads/"
+    ), "Do not include 'refs/heads/' in the remote branch name."
+
+    # List of hashes for commit on the local branch that are not on the remote branch
+    new_commit_local_hashes: list[str] = get_local_commits_not_on_remote(
+        local_branch_name, remote_branch_name
+    )
```

Comment:
great question - I candidly didn't think this through a ton. I re-read the script and think that the current behavior here is kinda bad; it'd likely just re-write files and not handle conflicts with the new changes safely at all. definitely need to re-architect that to see what's the better approach. 

do you have input here on what the desired behavior **_should_** be? I'm wondering about:
1. see if we could cleanly rebase our new commits onto the updated remote. if we cannot cleanly rebase onto the remote, then abort. 
2. if we can cleanly rebase onto the remote, then proceed, and push each commit one by one. If we see a new commit appear on the remote before all the mutations are done, bail when we see that - this would result in **_some_** of the commits being pushed to the remote

you've worked with some more of these usecases (Eg., translations) closely, so I'm curious what you think of this appraoch.


We could go the safer route and bail entirely if we diverge from the remote.

Reply:
the desired behavior IMO should be to immediately bail. i could see it being configurable instead, but i think we should wait until there's a relevant use-case.

usually for machine contexts, you're running assuming you're on the latest version, and that if you're not you'll get triggered again to do the same work.

Reply:
that sounds reasonable to me. it's also easier to implement, heh, so why not. if you don't know of any existing use-cases for this then bailing sounds perfect.